graphbolt/src/cuda/extension/gpu_graph_cache.cu

```diff
@@ -0,0 +1,322 @@
+            num_nodes,
+            kDoubleLoadFactor,
+            cuco::empty_key{static_cast<index_t>(-1)},
+            cuco::empty_value{std::numeric_limits<index_t>::lowest()},
+            {},
+            probing_t<index_t>{},
+            {},
+            {},
+            allocator_t<index_t>{},
+            cuco::cuda_stream_ref{cuda::GetCurrentStream()}};
+        map_ = new map_t<index_t>{std::move(map_temp)};
+      }));
+  C10_CUDA_KERNEL_LAUNCH_CHECK();  // Check the map constructor's success.
+  const auto options = torch::TensorOptions().device(c10::DeviceType::CUDA);
+  threshold_ = threshold;
+  device_id_ = cuda::GetCurrentStream().device_index();
+  map_size_ = 0;
+  num_nodes_ = 0;
+  num_edges_ = 0;
+  indptr_ = torch::zeros(num_nodes + 1, options.dtype(indptr_dtype));
+  for (auto dtype : dtypes) {
+    cached_edge_tensors_.push_back(
+        torch::empty(num_edges, options.dtype(dtype)));
+  }
+}
+
+GpuGraphCache::~GpuGraphCache() {
+  AT_DISPATCH_INDEX_TYPES(
+      cached_edge_tensors_.at(0).scalar_type(), "GpuGraphCache::GpuGraphCache",
+      ([&] { delete reinterpret_cast<map_t<index_t>*>(map_); }));
+}
+
+struct AddTuple {
+  template <typename T>
+  __host__ __device__ auto operator()(const T& a, const T& b) const {
+    return thrust::make_tuple(
+        thrust::get<0>(a) + thrust::get<0>(b),
+        thrust::get<1>(a) + thrust::get<1>(b));
+  }
+};
+
+std::tuple<
+    torch::Tensor, std::vector<torch::Tensor>, torch::Tensor, torch::Tensor,
+    torch::Tensor, int64_t>
+GpuGraphCache::Query(torch::Tensor seeds) {
+  auto allocator = cuda::GetAllocator();
+  auto index_dtype = cached_edge_tensors_.at(0).scalar_type();
+  const dim3 block(kIntBlockSize);
+  const dim3 grid((seeds.size(0) + kIntBlockSize - 1) / kIntBlockSize);
+  return AT_DISPATCH_INDEX_TYPES(
+      index_dtype, "GpuGraphCache::Query", ([&] {
+        auto map = reinterpret_cast<map_t<index_t>*>(map_);
+        while ((
+            map_size_ + seeds.size(0) >= map->capacity() * kDoubleLoadFactor)) {
+          map->rehash_async(
+              map->capacity() * kIntGrowthFactor,
+              cuco::cuda_stream_ref{cuda::GetCurrentStream()});
+        }
+        auto positions =
+            torch::empty(seeds.size(0), seeds.options().dtype(index_dtype));
+        CUDA_KERNEL_CALL(
+            _QueryAndIncrement, grid, block, 0,
+            static_cast<int64_t>(seeds.size(0)),
+            static_cast<index_t>(threshold_), seeds.data_ptr<index_t>(),
+            positions.data_ptr<index_t>(), map->ref(cuco::insert_and_find));
+        auto num_cache_enter =
+            allocator.AllocateStorage<thrust::tuple<int64_t, int64_t>>(1);
+        // Since threshold_ is a class member, we want the lambda functions
+        // below to only capture this particular variable by reassigning it to a
+        // local variable.
+        const auto threshold = -threshold_;
+        auto is_threshold = thrust::make_transform_iterator(
```

Comment:
is_threshold includes both is_threshold and is_new? can you come up with a better name?

Reply:
Changed to is_threshold_new_hit